Add ChainstateManager, remove BlockManager global

jamesob opened the pull request:
This is part of the [assumeutxo project](https://github.com/bitcoin/bitcoin/projects/11):

Parent PR: #15606 
Issue: #15605 
Specification: https://github.com/jamesob/assumeutxo-docs/tree/master/proposal

---

This changeset introduces `ChainstateManager`, which is responsible for creating and managing access to multiple chainstates. Until we allow chainstate creation from UTXO snapshots (next assumeutxo PR?) it's basically unnecessary, but it is a prerequisite for background IBD support.

Changes are also made to the initialization process to make use of `g_chainman` and thus clear the way for multiple chainstates being loaded on startup.

One immediate benefit of this change is that we no longer have the `g_blockman` global, but instead have the ChainstateManager inject a reference of its shared BlockManager into any chainstate it creates.

Another immediate benefit is that uses of `ChainActive()` and `ChainstateActive()` are now covered by lock annotations. Because use of `g_chainman` is annotated to require cs_main, these two functions subsequently follow.

Because of whitespace changes, this diff looks bigger than it is. E.g., 4813167d98 is most easily reviewed with
```sh
git show --color-moved=dimmed_zebra -w 4813167d98
```

jamesob committed:
validation: add CChainState.m_from_snapshot_blockhash

This parameter is unused, but in future commits will allow ChainstateManager to
differentiate between chainstates created from a UTXO snapshot from those that
weren't.

jamesob committed:
validation: introduce unused ChainstateManager

ChainstateManager is responsible for creating and managing multiple
chainstates, and will provide a high-level interface for accessing the
appropriate chainstate based upon a certain use.

Incorporates feedback from Marco Falke. Additional documentation written
by Russ Yanofsky.

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>

jamesob committed:
refactor: move RewindBlockIndex to CChainState

This is in preparation for multiple chainstate initialization in init.

jamesob committed:
use ChainstateManager to initialize chainstate

This allows us to easily initialize multiple chainstates on startup in future
commits. It retires the g_chainstate global in lieu of g_chainman.

jamesob committed:
test: add basic tests for ChainstateManager

Feedback incorporated from Russell Yanofsky.

Co-authored-by: MarcoFalke <falke.marco@gmail.com>

MarcoFalke commented on src/validation.h:
```diff
@@ -757,6 +762,111 @@ bool InvalidateBlock(BlockValidationState& state, const CChainParams& chainparam
+    uint256 SnapshotBlockhash() const { return m_snapshot_blockhash; }
+
+    //! Is there a snapshot in use and has it been fully validated?
+    bool IsSnapshotValidated() const { return m_snapshot_validated; }
```
in commit 6a629cfd029071: 

How and why is this different from `return !m_ibd_chainstate`?

jamesob replied:
See my comment above for background: https://github.com/bitcoin/bitcoin/pull/17737#discussion_r358858324. Perhaps we can do this without necessarily needing to maintain extra state (`m_snapshot_validated`) for it. I'll look into it.

MarcoFalke commented on src/validation.h:
```diff
@@ -757,6 +762,111 @@ bool InvalidateBlock(BlockValidationState& state, const CChainParams& chainparam
+
+    bool IsSnapshotActive() const;
+
+    uint256 SnapshotBlockhash() const { return m_snapshot_blockhash; }
```
in commit 6a629cfd0290716:

How and why is this different from `return m_active_chainstate->m_snapshot_blockhash;`?

jamesob replied:
I think you're right here, I'll remove this.

maflcko reviewed (approved):
ACK 0226408999, did not read the tests 🔲

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 0226408999, did not read the tests 🔲
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYEACgkQzit1aX5p
pUixgQv9ELRmdepH+VtXfc0mvkvHjZZAGrvdNEAGLqlEK+nbDy0YtUnZtECto3aJ
K/fFg7heGCu3/pRgx25GGDnaRNrOPc+q56kWXUjSS7lweiyx1JYw4V8sVlgas3Ih
4Cxbgy3GxmPL88e1mRBpqBdyKxGi3cPxpPjew63/sllo9ZDPBMsA5g0NcpUwIC2H
TMgdLJbffnHCgCKkIH5/lXah8W7fF6Tt/wyzK4RaijwcxjVrdFheJTv0kLiBJDAG
AmHMVhKDQTBeZFn5CydqeupqV90u7rdFd/S6MliuawQGwtlobeqKqs7jxeR23+Yx
AfAAk6nKLhQFQ5x998j/rGamIcYO7RCRE0be0dZC1KGdhTwIY1LOsSKHlJAn64Ik
XKNJXDbg8o1vnBgAEmzslunwooaOoF2oSyEB4crt++JNYLl38imrOdZ+0kdlQ/BC
GCRJWD/y2WSNRxCrkyJZ6uxTkUf7KmEbh8qwOavXby2DR2rOSOGE9nM9ewXgqbsK
BDO3176X
=Ygdp
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `c4bdc7880969f1e6719cf25c6f715a20e103935d1147395578a097e3179189ff  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108c4bdc7880969f1e6719cf25c6f715a20e103935d1147395578a097e3179189fff010bb951b70cdb0bcced5124390808ed35908fff010938af61d12bba8c46cb9f801b12d4cd908f1045df3bc8cf0080dbe291ea74b3c300083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010c9529e5b2eca39dac00f498a11a8d7f108f02076e7566f3ee1b0010b61c19def49d3dbeec9edb3ab706b6891cbec28c36bd01b08f1045df3bc8df00849d19536f3f49f560083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010cdab75cb221153d52136b1ea576296cb08f1045df3bc8cf008867deb40ab9aa08f0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010e063a6739f3ca420ca555b9ebe82a0aa08f1045df3bc8cf0080d40727b891495b30083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

promag commented on src/validation.h:
```diff
@@ -757,6 +762,111 @@ bool InvalidateBlock(BlockValidationState& state, const CChainParams& chainparam
+        bool activate = true, const uint256& snapshot_blockhash = uint256());
+
+    //! Get all chainstates currently being used.
+    std::vector<CChainState*> GetAll();
```
Agree. Same question for other methods.

jamesob replied:
Thanks for raising this point. Initially, I had all pointers in the ChainstateManager protected by a private `m_cs` lock. I later found that there were irreconcilable lock inversions that happened during initialization - because a function like `ChainActive()` would necessarily have to acquire `ChainstateManager.m_cs`, there were many tricky inversions that happened between that lock and others like `cs_main` since ChainActive() is used so pervasively throughout the code. 

As a result, I removed the chainman-specific lock and replaced it with `cs_main` in a later commit (https://github.com/bitcoin/bitcoin/pull/15606/commits/f1fa917e2081dd8d1847cbb23a0eae5a2cded2e5). Because ChainActive() isn't formally annotated as being protected by cs_main, and use of ChainstateManager always happens while under the protection of cs_main, I figured I could forgo adding those annotations now. But now that the point has been raised, I think it's probably worth addressing in this PR. 

So I'll add the `cs_main` annotations to the relevant ChainManager parts. Though of course this will entail annotating ChainActive() and there will come with many other (mostly trivial) changes.

jamesob replied:
Also, it's worth noting that the comment you quote is out of date (my fault!). We don't actually free the contents of the background validation chainstate *immediately* when validation completes - [we do it during shutdown](https://github.com/bitcoin/bitcoin/pull/15606/commits/8d6e0c34c32c5b95c126a963b16b5aebf707acbd#diff-24efdb00bfbe56b140fb006b562cc70bR5653-R5679) when `g_chainman.Reset()` is called. 

The reason I don't do it immediately after validation is complete is so that we can be sure that no other parts of the code are still making reference to that pointer. This also relates to your question below about `IsSnapshotValidated()` - the extra piece of state (`m_snapshot_validated`) is used to track whether or not we need to do this cleanup.

ariard commented on src/validation.cpp:
```diff
@@ -5116,3 +5116,68 @@ class CMainCleanup
+{
+    m_ibd_chainstate.reset();
+    m_snapshot_chainstate.reset();
+    m_active_chainstate = nullptr;
```
6a629cf

Maybe reset `m_snapshot_blockhash` to null and `m_snapshot_validated` to false?

ariard commented on src/init.cpp:
```diff
@@ -1566,7 +1566,7 @@ bool AppInitMain(NodeContext& node)
                 // BlockIndex() based on lack of available witness data.
                 uiInterface.InitMessage(_("Rewinding blocks...").translated);
-                if (!RewindBlockIndex(chainparams)) {
+                if (!::ChainstateActive().RewindBlockIndex(chainparams)) {
```
175e570

I'm not sure if `RewindBlockIndex` is still relevant.

Following Suhas comment https://github.com/bitcoin/bitcoin/pull/8149#issuecomment-257943026, it was added to let upgrading nodes after segwit activation redownload consensus data without redownloading the whole blockchain.

According to sipa in #15402, the rewinding logic "is probably becoming increasingly unnecessary, as very few pre-0.13.1 nodes remain that would care to upgrade".

Further, it doesn't make sense for the assumed-valid UTXO chainstate to rewind until first insufficiently-validated block, all utxo hash are going to be post segwit. It only makes sense for the ibd chainstate.

jamesob replied:
Agree this is worth investigating, but probably not within this PR. I'd rather just preserve existing behavior.

ariard replied:
FYI, https://github.com/bitcoin/bitcoin/issues/17862. I had a second look, I think it's safe to maintain logic as it is for this PR even it should be never triggered because both chainstates will belong to segwit nodes.

ariard commented on src/init.cpp:
```diff
@@ -1517,93 +1520,125 @@ bool AppInitMain(NodeContext& node)
-                                  gArgs.GetArg("-checkblocks", DEFAULT_CHECKBLOCKS))) {
-                        strLoadError = _("Corrupted block database detected").translated;
-                        break;
+                for (CChainState* chainstate : g_chainman.GetAll()) {
```
4813167

Just iterate for the `VerifyDB`, pruning and best block seen by RPC maybe should be common to both chainstates ?

jamesob replied:
Good point - I'm changing this to only report a new tip to RPC for the active chain. The tip/time check is very cheap and doesn't hurt to be done for both. There isn't anything that actually happens for pruning here other than a possible log statement, so I think leaving that as-is should be fine. Thanks!

ariard reviewed (commented):
Reviewed 6d14916 to 4813167.

Need to dig into parent PR a bit more before to review forward.

jamesob commented on src/validation.cpp:
```diff
@@ -5116,3 +5116,68 @@ class CMainCleanup
+
+    if (activate) {
+        LogPrintf("Switching active chainstate to %s\n", snapshot_blockhash.ToString());
+        m_active_chainstate = to_modify.get();
```
In practice all usages of this happen under `cs_main`, but as I said above I'll do explicit annotations.

jamesob commented on src/validation.h:
```diff
@@ -757,6 +762,111 @@ bool InvalidateBlock(BlockValidationState& state, const CChainParams& chainparam
+    //! @param[in] snapshot_blockhash   If given, signify that this chainstate
+    //!                                 is based on a snapshot.
+    CChainState& InitializeChainstate(
+        bool activate = true, const uint256& snapshot_blockhash = uint256());
```
I think this method is called so infrequently and has such major effects that it'd be hard to accidentally do anything one way or the other, and the results would be immediately obvious and buggy. But to your point, maybe I should remove the default parameters so that everything has to be explicit.

jamesob replied:
Removed the default parameters, thanks.

jamesob commented on src/init.cpp:
```diff
@@ -1463,8 +1467,7 @@ bool AppInitMain(NodeContext& node)
             bool is_coinsview_empty;
             try {
                 LOCK(cs_main);
-                // This statement makes ::ChainstateActive() usable.
```
I think it was incorrect since we haven't initialized the coins views at that point.

jamesob commented on src/validation.h:
```diff
@@ -757,6 +762,111 @@ bool InvalidateBlock(BlockValidationState& state, const CChainParams& chainparam
+ * *IBD chainstate*: a chainstate populated via initial block download.
+ *
+ * *Snapshot chainstate*: a chainstate populated by loading in an
+ *    assumed-valid UTXO snapshot.
```
Fixed, thanks.

jamesob commented on src/validation.cpp:
```diff
@@ -5116,3 +5116,68 @@ class CMainCleanup
+    to_modify.reset(new CChainState(snapshot_blockhash));
+
+    if (activate) {
+        LogPrintf("Switching active chainstate to %s\n", snapshot_blockhash.ToString());
```
Fixed, thanks.

jonatack commented:
Bringing in [this IRC comment](http://www.erisian.com.au/bitcoin-core-dev/log-2020-01-16.html#l-454), if helpful:

"recent utxo snapshots for testing are available here: https://github.com/bitcoin/bitcoin/pull/15606#issuecomment-555213164"

MarcoFalke commented on src/validation.cpp:
```diff
@@ -5139,3 +5147,82 @@ class CMainCleanup
+
+    if (activate) {
+        LogPrintf("Switching active chainstate to %s\n", to_modify->ToString());
+        m_active_chainstate = to_modify.get();
```
in commit 16d20d91b43b1c30333fe14272d69cfb543229b3:

You mention that "// If a snapshot chainstate exists, it will always be our active."

However, then you provide code here to activate a ibd chainstate even though a snapshot chainstate might exist. This is confusing and the code should probably not provide such a backdoor.

MarcoFalke replied:
Maybe a check like `if (m_active_chainstate && to_modify != m_snapshot_chainstate) LogPrintf("Error!"); else ...` or similar could prevent that?

jamesob replied:
Yep, this is a really good point. Fixed it to only allow a single activation sequence.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -5139,3 +5147,82 @@ class CMainCleanup
+
+CChain& ChainstateManager::ActiveChain() const
+{
+    return m_active_chainstate->m_chain;
```
in commit 16d20d91b43b1c30333fe14272d69cfb543229b3:

Should `assert m_active_chainstate` first?

MarcoFalke commented on src/validation.h:
```diff
@@ -775,6 +782,111 @@ bool InvalidateBlock(BlockValidationState& state, const CChainParams& chainparam
+    std::unique_ptr<CChainState> m_ibd_chainstate;
+
+    //! A chainstate initialized on the basis of a UTXO snapshot. If this is
+    //! non-null, it is always our active chainstate unless proven invalid.
```
in commit 16d20d9:

"proven invalid" sounds odd. In normal operation (with the assumevalid hash baked in) an assumevalid chainstate can never be invalid. And if it is, there is no way to recover from this, right? So proving that one or the other chainstate is invalid is impossible with conflicting information. The only way to handle this is to crash the node immediately. I'd say to remove the "unless proven invalid" phrase here.

jamesob replied:
Yep, good point.

MarcoFalke commented:
ACK 16d20d91b43b1c30333fe14272d69cfb543229b3 only, will continue review later on 🚀

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 16d20d91b43b1c30333fe14272d69cfb543229b3 only, will continue review later on 🚀
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjn6wv/cGwiv/yWquRmx2FEjuO62k8716JXtWFhx7zE1OcbgSYtOx2i4QAPQJhJ
FYhRtUuVArBrkz9lrn9VSCo3Aueb7w8b/j3/8dGyH/zqUU7IhWgrjC7qlfbdP4lp
DzDGE5ITqlvZlEkYAloI3zMoVRaUHDlGTK+bEHK3WdRnsPL9oR/C+S1UNcPNqq1m
I4G5UrneAHnQnIiWJfjfzpZ0BlXCXCfs3S4BBwKNTQAMlqlgzz1sCQr5R2d8kcW3
mNexZ76vn3IlrmFAyXWNwzQvGvfrGMF4RcIYQiS5OLfcjlzLfZOxSQJjerHzMe7C
EdnUsD+OX6BA6ijgU7eO7k5mdRxSkUVLanAzINUB5dupiw1NscUyZlqsH2J549DT
Q2g6va4jFXAULCFg2Rwx18k3GOQ0spLsRyv6wn/PxpwzEaZnEnSsoCe8pdiwCaOk
SyIJzqx3vzfA3IhEgHVnLHqLN8iYk9F5XXevDDA7fPN1A/KAWDcQXjVEpu3jA03w
nuOXSfZ+
=G8F6
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `7e94e353f869ccf97b1947f28a0c0193a5e2cf8d984598592f1296d23f14168d  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401087e94e353f869ccf97b1947f28a0c0193a5e2cf8d984598592f1296d23f14168df010932f6a93389aee9f66370eb628789b3008fff01004dcbe7b06025934db649ab4347e89ef08f1045e2b516cf008ede39c0dfa042ee40083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0100b1cd423e577a643596a0db531309fae08f1045e2b516df00861dda22d00ab11700083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010b968296e10819ea09540a39ea357315508f1045e2b516df00827bbb6020257f1af0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010e29ce239459275631a2919133ddc8ea308f1205b9e10df70f3727651768e1846a82711418f1e6299509dde211d4887f2f4162808f1045e2b516cf008882a6ad733fe9ab80083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

MarcoFalke commented on src/init.cpp:
```diff
@@ -1519,93 +1522,125 @@ bool AppInitMain(NodeContext& node)
+                    chainstate->InitCoinsCache();
+                    assert(chainstate->CanFlushToDisk());
+
+                    is_coinsview_empty = fReset || fReindexChainState ||
```
in commit e89fbd3cf0820ee263a94c49751a97a784acb049:

I have a hard time following the logic-change.

* You replace all calls to ChainActive with a for loop over all chainstates
* You only initialize the "ibd" chainstate, which will retain all logic as it was before
* However, is_coinsview_empty will be set to the result of whatever chainstate was processed last, and then used (in the next loop) on whatever chainstate is processed first. This seems fragile or at least requires a comment to explain why this can be done.

jamesob replied:
Maybe it's a little hard to see in the diff, but the current value of `is_coinsview_empty` is only ever used within a single loop. In the next loop, the [value is reset](https://github.com/jamesob/bitcoin/blob/1c173812bac34e263ae49e0bc80e45b437e95e8b/src/init.cpp#L1619-L1620) based on the chainstate being worked on.

It's safe to reset the value on line 1619 because it's not possible for the coinstip's best block to be set to null between this `is_coinsview_empty = ...` assignment and the one in the next loop.

To be extra sure there's no change, I could create a `chainstate_to_coinsview_empty` map and only set the values once, but I'll only do this if you think it's worthwhile.

MarcoFalke replied:
Oh, I missed that the value is reset. Maybe it would help to limit the scope of the variable to make that more clear and avoid this issue accidentally coming up in the future?

jamesob replied:
Fixed by removing the variable entirely and replacing it with a function.

MarcoFalke commented on src/init.cpp:
```diff
@@ -1519,93 +1522,125 @@ bool AppInitMain(NodeContext& node)
+                        strLoadError = _(
+                            "Unable to rewind the database to a pre-fork state. "
+                            "You will need to redownload the blockchain").translated;
+                        break;
```
in commit e89fbd3cf0820ee263a94c49751a97a784acb049:

Previously we'd break out, now we continue. Why?

jamesob replied:
Good catch! Fixed.

maflcko reviewed (commented):
ACK e7bc059491d51919f676b35b6843d1c395d452a6 only, have not reviewed later commits closely 🕧

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK e7bc059491d51919f676b35b6843d1c395d452a6 only, have not reviewed later commits closely 🕧
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjKXwwApeBAJd4ix/yrfHo9bHi30caJGej5Xk3cCvsbTorNmxVgMjJdbcALeADr
AbWHnVntgHiI7oLtzbKrFlKTGi826+stmRIy2VPyVs31L/z7k2pLKJsRFtYO+kzv
ATQ4POaLd58wzFI4bxGtSLBu7I+91f6iLzdtdINYDVJj6g89eBd74ICseFoQjM6x
lAugPzMXzrL8kh71IFBumNpQKq3rNZ9Wb/6oqVCBiZUGUnW4qC3a5aVDhwfnBnKX
YQia7Cej4Ex22+FAxh+4LUl8nO6I+UMH53TASaMG/ZoHUrXpHouWyo9vmUAwBMyp
gqN9AX/V5Hv7YCk3hehviw6p9ZI+Gvs3hrXYlU2FLUFTGXJfQz5yM6jSipnK5tZC
vqx46MazAjQ7x2de/HJIfwf+uEFpZkohauhobu5pMGss0tuVSiWVcofIoS8I5zei
mlEZ6atiGucuEMZmiC8+z3g4TbGo5kztOB00YOVVbY1+jDmuOOs+T220p4FtrDTl
B6J3CRmr
=Ik+8
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `fe215ba014f43a48217fc4a11afdd4fb671d3ccd7e48f8bed366d49affae1228  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108fe215ba014f43a48217fc4a11afdd4fb671d3ccd7e48f8bed366d49affae1228f01007ea2e6b65d1e638cc52d3cbbf197d5608fff0102b48a7a8d2242ef0b4089eea3e66b71e08f1045e3058adf008fe3c26b7d3ea59a40083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010d10fb3a95b08b3111763c1dc0d37ae4b08f1045e3058adf008b3976f4ae6099b560083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010dde2ae2e7272f0824fecf3e8fc513ae808f1045e3058adf008ba84d95521ba13ef0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented on src/init.cpp:
```diff
@@ -1519,93 +1522,125 @@ bool AppInitMain(NodeContext& node)
+                        }
+
+                        if (!CVerifyDB().VerifyDB(
+                                chainparams, &chainstate->CoinsDB(),
```
in commit e89fbd3:

VerifyDB will verify the db based on the active chain. How is this supposed to work when you pass in a db that is not from the active chainstate?

jamesob replied:
Oh good point - looks like I split the commits up a bit out of order. I could pull this (https://github.com/bitcoin/bitcoin/pull/15606/commits/23a074b6630dd75fd4a4a5b6495fe046012f2cc9) in or only do verification for the active chain... any preference?

MarcoFalke replied:
If you think that the commit that switches `AppInitMain` to use for-loops is blowing up this pull request too much, you can also leave it for later, I guess. I haven't checked, but it should be correct to just call `g_chainman.InitializeChainstate();` to initialize the "ibd" chainstate and then leave all calls as they were previously, i.e. `ChainstateActive()....`

MarcoFalke commented on src/init.cpp:
```diff
@@ -701,11 +701,21 @@ static void ThreadImport(std::vector<fs::path> vImportFiles)
-        StartShutdown();
-        return;
+
+    // XXX this is kind of a hack to avoid a lock inversion.
```
in commit 1c173812bac34e263ae49e0bc80e45b437e95e8b:

Why is this a hack and what should be used instead of the hack? If there is nothing that can be used instead, this is not a hack and the comment should go.

jamesob replied:
Fixed by cleaning up inline comments and adding documentation to the ChainstateManager pointer definitions.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1403,7 +1403,15 @@ void static ProcessGetBlockData(CNode* pfrom, const CChainParams& chainparams, c
         BlockValidationState state;
-        if (!ActivateBestChain(state, Params(), a_recent_block)) {
+
+        // XXX this is kind of a hack here to get around the fact that we can't
```
Same here and everywhere else you use the `XXX`

MarcoFalke commented:
ACK 1c173812bac34e263ae49e0bc80e45b437e95e8b, but it looks like the tests fail 🍢

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 1c173812bac34e263ae49e0bc80e45b437e95e8b, but it looks like the tests fail 🍢
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjFNgwApRAQI5N2D4113l6z1rLjgUjlixBjaHFnZC0rG5naAL/Gu55ofeAmzg5Z
TYF1WvqL9spMC+CDCweOyRGKsWb81Ryv8Ieni+Dn8RE+pu0VLakDu9Lw95+bd43j
1P6RLJYa2NYFlCuJssbs5ziEPJqs3/DlelgWPslfbdgwLXWFT54CvijXuFEF43b/
T9/NBr1OO1ByA0+AEM0eT3iHH55Guna544fuY8UjEDabnjqs8EiatmXArGlqi6Wr
D6U7J33USdtpof1bb+2zR8KLSb4MCFFxezr2yrSREHf5tqqamfN3qJ2dYHazNEMd
S4ka7rOrQqY/c2XIBs4oYD8pxI3thvxuo4nAnsG0sxjxkW0mm4O5OBqTFP/aGilX
NNkR1TTl1DcuqbTFK3FNoiqmG/yuh2OWw3jm0EG6jhEcf3sdf1d2e5RUQ0xKPDRm
h/PVqKPlmRDkEHiT9XoLgWDw6C9fH86HdGh0HLBZmTJ2jOt+sVeIvSXtth7sVznG
vtfM49Wg
=0mLY
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `6a38ed3fc7e109b015b25d8cd883d76b0d7c7f56f8c96dd80e10381889e51bbd  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401086a38ed3fc7e109b015b25d8cd883d76b0d7c7f56f8c96dd80e10381889e51bbdf010aa5fff91763b0cf471495be301eed42608fff0101ef862cc09f5549b6fe9bd6be7aa9ab308f1045e3084fef00840dc1af3f0f867230083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff01060445eda4658bc26a60a7fb78cea975908f1045e3084fdf00843441668b950cb840083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010a59c899827ea8fb57697330aa2dde3cb08f1045e3084fef008291d3bc5bd7b02730083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010d1af961bfb84c92bb1d87fe19689d14e08f1045e3084fdf008ce3f6b62ea653e220083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

ryanofsky reviewed (approved):
Code review ACK 1c173812bac34e263ae49e0bc80e45b437e95e8b. I did have a bunch of questions and suggestions below, but the suggestions aren't important, and if the questions don't point to real problems, I think this looks good to merge as is.

jamesob commented on src/validation.h:
```diff
@@ -775,6 +782,109 @@ bool InvalidateBlock(BlockValidationState& state, const CChainParams& chainparam
+ *
+ * *Background validation chainstate*: an IBD chainstate for which the
+ *    IBD process is happening in the background while use of the
+ *    active chainstate allows the rest of the system to function.
```
What gets kind of confusing here is that chainstates created from snapshots still technically go through IBD, since by the time the snapshot is actually used to create a chainstate, it's far enough behind the tip that it technically undergoes IBD (albeit an abbreviated one). So I think referring to what I now call the background validation chainstate as the IBD chainstate might be confusing in that sense.

But I agree with you that the naming here isn't ideal. Definitely open to changing this if you have other good ideas.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/17737#discussion_r372099708

I guess I see that but it would seem more like a reason to avoid the term "IBD chainstate" entirely. 

Since chain state manager only has a snapshot chainstate and a non snapshot chainstate, I like the idea of just referring to the chainstates with the same names consistently, and not muddying the water with other chainstate names and definitions. "IBD chainstate" is what you call the non-snapshot chainstate everywhere except in this one method and comment, which is my why first suggestion is delete "background validation chainstate" comment and have separate methods replacing IsBackgroundValidationChainstate. Switching to "Background IBD chainstate" was just a runner up suggestion.

I guess also I don't find the "IBD chainstate" name to be too confusing because I take IBD chainstate to mean chainstate generated entirely from IBD, not a state with some UTXOs that happened to come from IBD. But for another naming suggestion, maybe an alternative would be to use "snapshot chainstate" and "blocksonly chainstate" instead of "snapshot chainstate" and "IBD chainstate"

jamesob replied:
Partially (?) fixed by changing `IsBackgroundValidationChainstate()` to `IsBackgroundIBD()`. I think blocksonly is somewhat confusing because there'd be a confusing overlap with the P2P definition of blocksonly, which I think is different or unrelated.

jamesob committed:
protect g_chainman with cs_main

I'd previously attempted to create a specialized lock for ChainstateManager,
but it turns out that because that lock would be required for functions like
ChainActive() and ChainstateActive(), it created irreconcilable lock inversions
since those functions are used so broadly throughout the codebase.

Instead, I'm just using cs_main to protect the contents of g_chainman.

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>

jamesob commented:
[`au.chainman.6`](https://github.com/jamesob/bitcoin/tree/au.chainman.6) -> [`au.chainman.9`](https://github.com/jamesob/bitcoin/tree/au.chainman.9) [(changes)](https://github.com/bitcoin/bitcoin/compare/1c173812bac34e263ae49e0bc80e45b437e95e8b..94ca1e07bb56a1d1d4d21871056144ae00941ef3)

I've incorporated most if not all of the feedback from @MarcoFalke and @ryanofsky which has nicely reduced the size of the diff. Changes include:
- fixing variable scoping, control flow, omissions in the `init` changes
- acquiring `::cs_main` directly in the chainstate convenience functions (instead of having to annotate calls with `LOCK`)
- clarifying chainstate lifetimes in ChainstateManager documentation
- fixing the unittests for QT use

jamesob commented on src/validation.h:
```diff
@@ -757,6 +762,111 @@ bool InvalidateBlock(BlockValidationState& state, const CChainParams& chainparam
+ *    IBD process is happening in the background while use of the
+ *    active chainstate allows the rest of the system to function.
+ *
+ * *Validated chainstate*: a chainstate whose validity is not assumed.
```
Fixed, using Russ' definition.

jamesob commented on src/validation.h:
```diff
@@ -775,6 +782,109 @@ bool InvalidateBlock(BlockValidationState& state, const CChainParams& chainparam
 
+/**
+ * Provides an interface for creating and interacting with multiple
+ * chainstates, as well as a means to initialize chainstates from a
```
Fixed, thanks!

jamesob commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -0,0 +1,113 @@
+    for (size_t i = 0; i < v1.size(); ++i) {
+        BOOST_CHECK_EQUAL(v1[i], v2[i]);
+    }
+}
```
Fixed, thanks.

jamesob commented on src/init.cpp:
```diff
@@ -237,8 +237,10 @@ void Shutdown(NodeContext& node)
         LOCK(cs_main);
-        if (g_chainstate && g_chainstate->CanFlushToDisk()) {
-            g_chainstate->ForceFlushStateToDisk();
+        for (CChainState* chainstate : g_chainman.GetAll()) {
```
Good catches! Fixed, thanks.

jamesob commented on src/init.cpp:
```diff
@@ -1519,93 +1522,125 @@ bool AppInitMain(NodeContext& node)
+            for (CChainState* chainstate : g_chainman.GetAll()) {
+                if (!fReset) {
+                    // Note that RewindBlockIndex MUST run even if we're about to -reindex-chainstate.
+                    // It both disconnects blocks based on ::ChainActive(), and drops block data in
```
Fixed.

jamesob commented on src/validation.h:
```diff
@@ -891,16 +891,16 @@ class ChainstateManager
 
 /** @returns the most-work valid chainstate. */
-CChainState& ChainstateActive();
+CChainState& ChainstateActive() EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
```
Fixed by taking your suggestion. I didn't initially do this because I thought it created some kind of lock annotation problem, but after trying this change and rebasing the parent PR onto it, clang threw no warnings. Subsequently the diff has contracted a bit - thanks!

laanwj commented on src/validation.cpp:
```diff
@@ -5126,10 +5126,99 @@ class CMainCleanup
+        throw std::logic_error("unexpected chainstate activation");
+    }
+
+    return *to_modify.get();
```
`to_modify->get()`?

MarcoFalke replied:
Heh, member access has a higher precedence than indirection (dereference), so this is correct. Though, the same could be written shorter as just `*to_modify`

jamesob replied:
Went with Marco short-form.

jamesob commented:
[`au.chainman.9`](https://github.com/jamesob/bitcoin/tree/au.chainman.9) -> [`au.chainman.11`](https://github.com/jamesob/bitcoin/tree/au.chainman.11)

Pushed fixes for some small bugs, @MarcoFalke's `VerifyDB` comment, and [one big embarrassing typo](https://github.com/bitcoin/bitcoin/compare/94ca1e07bb56a1d1d4d21871056144ae00941ef3..ead4c5fada4246e1a37551469430a07c04d783a7).

ryanofsky commented on src/validation.cpp:
```diff
@@ -3818,7 +3821,7 @@ bool ProcessNewBlock(const CChainParams& chainparams, const std::shared_ptr<cons
 
     BlockValidationState state; // Only used to report errors, not invalidity - ignore it
-    if (!::ChainstateActive().ActivateBestChain(state, chainparams, pblock))
+    if (!ActivateBestChain(state, chainparams, pblock))
```
In commit "protect g_chainman with cs_main" (d88f75e495a46ae474d8ad67faaec13ef1950f5e)

I don't understand the two swaps between `::ActivateBestChain(...)` and `::ChainstateActive().ActivateBestChain(...)` in this commit. They don't seem to do anything, and if I revert them there don't seem to be obvious problems.

Not specific to this PR but generally it seems confusing to have an instance method and a global function both named `ActivateBestChain`. Would suggest renaming one or the other (this is the reason I like lowercase method names) or maybe just dropping the global function since it is only one line.

jamesob replied:
Yeah, this was churn from when I was reverting the inline cs_main locking I was initially doing for the last commit. Reverted, good catch. 

I'm for your ActivateBestChain suggestions, but would prefer to do them in a follow-up PR to limit the scope of this one. If anyone thinks that'd be best done here, I'm happy to.

ryanofsky commented on src/validation.h:
```diff
@@ -822,14 +822,41 @@ class ChainstateManager
     //! non-null, it is always our active chainstate.
+    //!
+    //! Once this pointer is set to a corresponding chainstate, it will not
+    //! be reset until init.cpp:Shutdown(). This means it is safe to acquire
```
In commit "protect g_chainman with cs_main" (d88f75e495a46ae474d8ad67faaec13ef1950f5e)

This doesn't explain what happens if you load two snapshots (one after the other). It seems like `ChainstateManager::InitializeChainstate` explicitly allows this so it would be good to add some clarification here or there.

jamesob replied:
I don't think `InitializeChainstate` allows this because of this check: https://github.com/jamesob/bitcoin/blob/8fe8b2e499ceb61c6fd3b73bf1c5dbc5861ccf62/src/validation.cpp#L5164-L5170

I wrote it to not allow the initialization of more than one "kind" of chainstate.

ryanofsky commented on src/init.cpp:
```diff
@@ -1519,93 +1521,129 @@ bool AppInitMain(NodeContext& node)
+
+                        // Only verify the DB of the active chainstate. This is fixed in later
+                        // work when we allow VerifyDB to be parameterized by chainstate.
+                        if (&::ChainstateActive() == chainstate &&
```
In commit "use ChainstateManager to initialize chainstate" (741db94e52b51d4df8173d56bfb1258ccf9ddbb3)

Just to understand, is checking ChainstateActive() == chainstate here an optimization to avoid verifying the same db multiple times? Could update comment, but no need to since hopefully this will be pretty shortlived

jamesob replied:
Yeah, it's just because `VerifyDB` at the moment makes explicit reference to `ChainActive`: https://github.com/jamesob/bitcoin/blob/8fe8b2e499ceb61c6fd3b73bf1c5dbc5861ccf62/src/validation.cpp#L4224-L4234

ryanofsky reviewed (approved):
Code review ACK d88f75e495a46ae474d8ad67faaec13ef1950f5e. Biggest change since last review is last commit which is rewritten to move locking inside validation.cpp accessors requiring locking by callers. Also there were some tweaks & minor fixes to the init code, some comment/renaming/test cleanups, and a qt test logic_error workaround

MarcoFalke commented on src/test/util/setup_common.cpp:
```diff
@@ -109,7 +109,18 @@ TestingSetup::TestingSetup(const std::string& chainName) : BasicTestingSetup(cha
+        if (msg.find("overwriting a chainstate") == std::string::npos) {
+            throw;
+        }
+    }
```
Please remove this hack in the unit tests. It will lead to potentially not resetting global state and preserving a chainstate of previous tests. The test failure is due to a missing reset in the gui tests, which can be fixed with a one-line patch:


```diff
diff --git a/src/qt/test/apptests.cpp b/src/qt/test/apptests.cpp
index 14a75b23f3..f9eb4cde30 100644
--- a/src/qt/test/apptests.cpp
+++ b/src/qt/test/apptests.cpp
@@ -82,6 +82,7 @@ void AppTests::appTests()
     // Reset global state to avoid interfering with later tests.
     AbortShutdown();
     UnloadBlockIndex();
+    g_chainman.Reset();
 }
 
 //! Entry point for BitcoinGUI tests.

jamesob replied:
Ah, thanks for the patch. Fixed.

MarcoFalke commented:
re-ACK d88f75e495a46ae474d8ad67faaec13ef1950f5e  will probably do a fresh review some time later 🍶

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK d88f75e495a46ae474d8ad67faaec13ef1950f5e  will probably do a fresh review some time later 🍶
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjJuQv9Fd31SrPN6kA/2g2p3Z99NXkVHa4CFGg9Eo+sCalKZ7YvQ3KwrKsC1AXK
uf9phsF6KBNX3cx7E9MEJ1OUCh5s3zJeBjZ3D6uK8xb9qvqA5ow87xb858h87isY
NRHXWjNYJHYORCitPhoK3hFws5lPxSf8K6K9FbbNISg4PY2zWEeaWGKmmxCPtRF2
ho6E5YSK/MyHVxXLUdiWQ3lpv0OEC9CustJ0KtSWOL4Vix5OVQJYlyxJTfnrE4Bn
I7axZLEQ6H0RhJWKha0/+3ntclZS4MC1dO6Qk12pQzxXl7K9tz5ppePXIR9/8dRt
SG38Ht92ff1XMq0ENIaMdiwn/XiH1b78ZZGztuXQuEAkUGj1w/eM+nBTVvSm8Ndj
Fu17KtFvE8lbDdKXRhdlvTHLpWEzzeaOkJQHslCMG+QpipzYHOvm4Iu173CHjXDi
PARCm2B9vQ3PDcvSD+kJxkbo53pSStv1uoM5zHyGRAXpjawmpB9PJpWuPdRTrGlx
V/E5uVAQ
=0Exq
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `8e3d2a68db66c60ea8e7ee3a4ff26522646e5da2bf7ceb11af7b3d8750f0973d  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401088e3d2a68db66c60ea8e7ee3a4ff26522646e5da2bf7ceb11af7b3d8750f0973df010c88e86e1b7a4d12cba7eb87a63051a8f08fff01035d4b44f6f5bb494c8ec6191c6fcfef708f1045e31d03ef0080a2d39aabb5059b60083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0104397b6a0acb073454363feb500acb7c608f1045e31d03df008d0f2f500acf7b4e60083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01063c0edb3771dac86cbc5a35f9fa137b008f1045e31d03ef008ea05635c59ff78900083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010e6e6e1b2100f851df27d2175d7d3bcf608f1045e31d03ef008d3c65b6e03d1c9a60083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

jamesob commented on src/validation.cpp:
```diff
@@ -5126,10 +5126,99 @@ class CMainCleanup
 };
 static CMainCleanup instance_of_cmaincleanup;
+
+uint256 ChainstateManager::SnapshotBlockhash() const {
```
Good point, fixed.

jamesob commented on src/init.cpp:
```diff
@@ -698,11 +698,20 @@ static void ThreadImport(std::vector<fs::path> vImportFiles)
+    std::vector<CChainState*> chainstates;
+    WITH_LOCK(::cs_main, chainstates = g_chainman.GetAll());
+
+    for (CChainState* chainstate : chainstates) {
```
Ah yeah, that's a nice improvement. Fixed.

jamesob commented:
[`au.chainman.11`](https://github.com/jamesob/bitcoin/tree/au.chainman.11) -> [`au.chainman.12`](https://github.com/jamesob/bitcoin/tree/au.chainman.12)[(changes)](https://github.com/bitcoin/bitcoin/compare/d88f75e495a46ae474d8ad67faaec13ef1950f5e..8fe8b2e499ceb61c6fd3b73bf1c5dbc5861ccf62)

Thanks for the continued reviews!

MarcoFalke commented:
Tests fail when compiled with `--enable-debug`

ryanofsky reviewed (approved):
Code review ACK 8fe8b2e499ceb61c6fd3b73bf1c5dbc5861ccf62. Changes since last review: making SnapshotBlockhash return optional, removing qt test workaround, minor style / include tweaks

Are there still problems with tests?

ryanofsky commented:
> Code review ACK [8fe8b2e](https://github.com/bitcoin/bitcoin/commit/8fe8b2e499ceb61c6fd3b73bf1c5dbc5861ccf62). Changes since last review: making SnapshotBlockhash return optional, removing qt test workaround, minor style / include tweaks
> 
> Are there still problems with tests?

Running this locally, I see a simple error in a wallet test that is locking `cs_main` in `ChainActive()` after `cs_wallet` is locked instead of before. Following fix works:

```diff
--- a/src/wallet/test/wallet_tests.cpp
+++ b/src/wallet/test/wallet_tests.cpp
@@ -456,7 +456,7 @@ public:
         CreateAndProcessBlock({}, GetScriptForRawPubKey(coinbaseKey.GetPubKey()));
         wallet = MakeUnique<CWallet>(m_chain.get(), WalletLocation(), WalletDatabase::CreateMock());
         {
-            LOCK(wallet->cs_wallet);
+            LOCK2(::cs_main, wallet->cs_wallet);
             wallet->SetLastBlockProcessed(::ChainActive().Height(), ::ChainActive().Tip()->GetBlockHash());
         }
         bool firstRun;
```

```c++
2020-02-07T21:36:34Z POTENTIAL DEADLOCK DETECTED
2020-02-07T21:36:34Z Previous lock order was:
2020-02-07T21:36:34Z  (1)
 cs_main interfaces/chain.cpp:239 (in thread )
2020-02-07T21:36:34Z  (2)
 cs_wallet wallet/wallet.cpp:2973 (in thread )
2020-02-07T21:36:34Z Current lock order is:
2020-02-07T21:36:34Z  (2)
 wallet->cs_wallet wallet/test/wallet_tests.cpp:459 (in thread )
2020-02-07T21:36:34Z  (1)
 ::cs_main validation.cpp:92 (in thread )
Assertion failed: detected inconsistent lock order at sync.cpp:119, details in debug log.
```

jamesob commented:
[`au.chainman.12`](https://github.com/jamesob/bitcoin/tree/au.chainman.12) -> [`au.chainman.13`](https://github.com/jamesob/bitcoin/tree/au.chainman.13) [(changes)](https://github.com/bitcoin/bitcoin/compare/8fe8b2e499ceb61c6fd3b73bf1c5dbc5861ccf62..abfc15273f817d30c3f53f910a8b340a72a6377a)

Thanks for the fix, @ryanofsky. Tests are now passing.

ryanofsky commented:
Code review ACK abfc15273f817d30c3f53f910a8b340a72a6377a. Changes since last review:

```diff
+Co-authored-by: Russell Yanofsky <russ@yanofsky.org>
```

with my test fix. Waiting for my github royalties to start rolling in.

promag reviewed (commented):
Code review ACK abfc15273f817d30c3f53f910a8b340a72a6377a

ariard commented on src/validation.h:
```diff
@@ -894,7 +921,7 @@ class ChainstateManager
 };
 
-extern ChainstateManager g_chainman;
+extern ChainstateManager g_chainman GUARDED_BY(::cs_main);
```
abfc152

If understand well the new lock model, `g_chainman` content is under `cs_main` but not `CChainState`. If yes, why some `GetAll` access in init.cpp don't use lock for access ?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/17737#discussion_r385962192

> If understand well the new lock model, `g_chainman` content is under `cs_main` but not `CChainState`. If yes, why some `GetAll` access in init.cpp don't use lock for access ?

I don't think this is true. Maybe it was resolved? You can add EXCLUSIVE_LOCKS_REQUIRED(cs_main) to the GetAll declaration, and it will only fail in test code using a local chain state manager not tied to the cs_main global

ariard reviewed (commented):
Mostly minor points, just need another parse on new lock implications to Code Review ACK.

ariard commented on src/validation.cpp:
```diff
@@ -5139,3 +5148,90 @@ class CMainCleanup
+}
+
+bool ChainstateManager::IsBackgroundIBD(CChainState* chainstate) const
+{
```
68998ec

Assert `m_snapshot_chainstate` == `m_active_chainstate` ?

fjahr commented on src/init.cpp:
```diff
@@ -698,11 +698,17 @@ static void ThreadImport(std::vector<fs::path> vImportFiles)
     }
 
     // scan for better chains in the block chain database, that are not yet connected in the active best chain
```
nit: That comment is still relevant to the code below but because of the empty line it seems detached.

fjahr reviewed (commented):
Code Review ACK abfc15273f817d30c3f53f910a8b340a72a6377a

Had a few nits that can well be ignored or addressed in next PR in the assume-utxo series.

General note: I was not sure about some of the `ChainstateManager` API, some of the methods seemed like they could be private (`IsSnapshotValidated()` seems to be only used internally) and others are not used at all (`IsSnapshotValidated` and `SnapshotBlockhash` for example) so they could have been included in the next PRs when they are needed. But I can wait and see how they are used in the next PR. :)

fjahr commented:
Code Review Re-ACK abfc15273f817d30c3f53f910a8b340a72a6377a

MarcoFalke commented:
re-ACK c9017ce3bc27665594c9d80f395780d40755bb22 📙

Only changes since my last review:

* Fix lock order in tests
* Making the snapshot blockhash optional


<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK c9017ce3bc27665594c9d80f395780d40755bb22 📙

Only changes since my last review:

* Fix lock order in tests
* Making the snapshot blockhash optional

-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjHLQv/RAhLnsTpzdzOpmyvTALXqbxQut9dguj1Tt46DsWnRzIcrwwcsYHbDreU
B3DmuqYHMdcb1o/OWV5IvWiHMnUbY/EbFfD6Zw9/4HNhHxmruCgUrG9+zP+oD1ck
nacI226jyJQWVKEz6XtbuF7uhW1an9J2XyPbnh6EaUywTB+CabLI/mnOFiHerWGd
Q4VwcxSFYeBjmoDP1ekDaTafAphHhDSDO7MaissHBPq+1HliMURndLGgDKpRnT0y
Fn7dfmN/fN3bS5JWufUCaatPdYjES5UOWRaDasub0ipmlvBXzTk/Z48lYzqz4QEp
Owa9fl6EoiRrhsQQSMOdEm6Mi2RCkwIXTK85+Ix7VP7ys+uQov/5fJbxel9zhInK
/4nAtO6uum9XscLqWqHi6AFEVGOEVOfYPN+CJbRW1y5PinpR0t1kRVhEjfXgzWag
bZfFWB2qs9bo+pWMCBDB/xeucJbJC2aDfnDqE5Gqpf1Knrg7t2tVbgU7IWJBaQLb
9KZNfpTL
=hs4D
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `f61dcf850f62e4c3ff5ca63de6c032928f672c45311925c6f66a2497f5bc0c48  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108f61dcf850f62e4c3ff5ca63de6c032928f672c45311925c6f66a2497f5bc0c48f010c95c01ec966b9e8599f5555e2d9c1cb508fff01053700317014be9c68812c12798aff0aa08f1045e723940f0088b8c7c3de3b398560083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0106682cfbba41201c9bbd28900494f9d9308f1045e723940f0080a2c3f72c53485980083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010a1e794bc9149f24614109cb95570960d08f1045e723940f00820cf08bc435f6ee10083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010b0d51ad27672195ef743b679a9b79b3208f1045e72393ff008bbb3fbea690636b40083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented:
@fjahr It looks like your commit hash is the same in the re-ACK

fjahr commented:
> @fjahr It looks like your commit hash is the same in the re-ACK

Ugh, how did I do that? I definitely re-reviewed the new code.

Code Review Re-ACK c9017ce3bc27665594c9d80f395780d40755bb22

ariard commented on src/validation.cpp:
```diff
@@ -81,12 +81,14 @@ ChainstateManager g_chainman;
 
 CChain& ChainActive()
 {
+    LOCK(::cs_main);
```
c9017ce

Is this LOCK really necessary given we call `::ChainstateActive()` just after ? Doesn't hurt because we use recursive_mutex but still..

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/17737#discussion_r395205730

> Is this LOCK really necessary given we call `::ChainstateActive()` just after ? Doesn't hurt because we use recursive_mutex but still..

Agree it's definitely not necessary (but harmless)

ariard reviewed (commented):
Code Review ACK c9017ce

Side-node: there is a `g_chainman` access in `LoadBlockIndex` (src/validation.cpp L4577), not an issue right now because caller already holds `cs_main` but if function is moved in the future a `AssertLockHeld` would be better.

ryanofsky reviewed (approved):
Code review ACK c9017ce3bc27665594c9d80f395780d40755bb22. No changes since last review other than a straight rebase

jamesob commented:
This tip has accumulated a good number of code ACKs, so I'm hesitant to address nits here. If anyone thinks that's the right move though I'm happy to.

fanquake commented:
>  If anyone thinks that's the right move though I'm happy to.

I think we can merge this after the 0.20 branch off.
